Name: Zahra Buchanan
Student #: 1011018818
Lecture: 0101

In [3]:
'''PRE-LECTURE Q1'''
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers = pd.read_csv(url)
villagers.isna().sum()

'''PRE-LECTURE Q2'''
# Get column names
columns = villagers.columns.tolist()

# Get the number of rows and columns
num_rows, num_columns = villagers.shape

columns, num_rows, num_columns

num_rows, num_columns = villagers.shape

print(f"Number of rows: {num_rows}")
print(f"Number of columns: {num_columns}")

'''PRE-LECTURE Q3'''
# Basic information about the dataset (data types, non-null counts)
villagers.info()

# Summary statistics for numerical columns
print(villagers.describe())

# Summary of all columns (including non-numerical)
print(villagers.describe(include='all'))

# 1. Count the number of villagers by species
species_counts = villagers['species'].value_counts()
print("Villagers by Species:")
print(species_counts)

# 2. Count the number of villagers by personality
personality_counts = villagers['personality'].value_counts()
print("\nVillagers by Personality:")
print(personality_counts)

# 3. Count the number of villagers by gender
gender_counts = villagers['gender'].value_counts()
print("\nVillagers by Gender:")
print(gender_counts)

# 4. Count the number of villagers by a specific column, sorted in ascending order
species_counts_sorted = villagers['species'].value_counts(ascending=True)
print("\nVillagers by Species (sorted in ascending order):")
print(species_counts_sorted)

'''PRE-LECTURE Q4'''
url2 = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic = pd.read_csv(url2)
titanic.isna().sum()

# Get the number of rows and columns
num_rows2, num_columns2 = titanic.shape

columns, num_rows2, num_columns2

num_rows2, num_columns2 = titanic.shape

print(f"Number of rows: {num_rows2}")
print(f"Number of columns: {num_columns2}")

titanic.describe()



Number of rows: 391
Number of columns: 11
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000
             row_n       id     name gender species birthday personality  \
count   391.000000      390   

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Q2.2 A variable refers to the various attributes that an entity can posesss and these make up the columns in a dataframe.
For example, in the villagers dataframe the columns species, gender and personality are examples of variables of the different entities.
An observation refers to an entire row in the dataframe inclusive of all variables related to that row. For example, an observation in the villagers dataframe would include the name, species, personality, gender etc. of an entity.

Q4. The difference with how the missing values are represented is that .shape does not take into consideration missing values and just reports how many rows and columns are in the data set. On the other hand, .describe only shows the columnss with numeric values and it includes a row, "count", which shows the number of all non-missing values in each column.

Q5. .shape is an attribute meaning it is meant to provide information on an object that already exists. In this case .shape retirieves the dimensions(# of rows and # of columns) of the dataframe without calling any functions or doing any calculations. Unlike .shape, .describe() is a method meaning it is considered a function which only works with certain objects/ data types. When .describe() is used a function is called and new information is calculated to be displayed such as the various columns(mean, count, std etc.). 

Chatbot link for Pre Lecture HW: https://chatgpt.com/share/c4d1b3ef-922e-46aa-b8e3-358fdc3a8683

POST LECTURE HW
Q6. count:shows the number of all non-missing values in each column
mean:shows the average value for each column(sum of everything divided by the number of non-missing values)
std: this shows the standard deviation of each column which shows the overall variation from the mean
min: shows the minimum value for each column exclusive of any missing values
25%: this refers to the median of the lower half of the data which is also the value where 25% of the data falls and is known as the first quartile(Q1)
50%: this is the median of each column which is also known as the second quartile(Q2)
75%: this refers to the median of the upper half of the data which is also the value where 75% of the data falls and is known as the third quartile(Q3)
max: this shows the maximum value for each column

Q7.1 A use case in which .dropna() is preferred would be if someone was handling a data set on movies and they were looking to analyze the ratings of each movie but they end up finding a few missing values in the 'ratings' column. Instead of using del.movies['ratings'] they would use movies.dropna() to remove all rows with missing values for the rating. This is used instead as deleting the entire column is not an option because the data in that column is needed for analysis.

7.2 A use case for the opposite could be if someone was handling a movie data set and wanted to analyze the ratings of the movies but found the dataset had a column titled 'budget'(showing the budget for each movie) in which many of the values were missing. Since the person will not need the data in that column to analyze the ratings they could use del.movies['budget'] to remove the entire column.

7.3 It is important to use del before .dropna() as its more efficient and can prevent loss of important data. Since .dropna() removes all rows with missing values, if there is an unnecessary column with mainly missing values using .dropna() will remove all the corresponding rows and majority of your data will be gone. If del is used first to remove the unnecessary column then using .dropna() will be sure to only remove the rows not useful for analysis.
7.4

In [10]:
import pandas as pd

url2 = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic = pd.read_csv(url2)
titanic.isna().sum()

# Get the number of rows and columns
num_rows2, num_columns2 = titanic.shape
num_rows2, num_columns2

num_rows2, num_columns2 = titanic.shape

print("Before using del titanic['age'] and titanic.dropna() ")
print(f"Number of rows: {num_rows2}")
print(f"Number of columns: {num_columns2}")

print (titanic.head())
print(" ")
titanic.describe()

'''POST LECTURE Q7.4'''
del titanic['age']
#using del to remove the age column since it is the only column with missing values
titanic.dropna()
#since no other columns have missing values .dropna() wont have any effect on the dataset

num_rows2, num_columns2 = titanic.shape

print("After using del titanic['age'] and titanic.dropna() ")
print(f"Number of rows: {num_rows2}")
print(f"Number of columns: {num_columns2}")
print(" ")

print (titanic.head())

'''POST-LECTURE Q8.1'''
titanic.groupby("sex")["sibsp"].describe()

Before using del titanic['age'] and titanic.dropna() 
Number of rows: 891
Number of columns: 15
   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
 
After using del titanic['age'] and titanic.dropna() 
Number of rows: 891
Number of columns: 14
 
   survived  pclass     

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.694268  1.156520  0.0  0.0  0.0  1.0  8.0
male    577.0  0.429809  1.061811  0.0  0.0  0.0  0.0  8.0

8.1 dfgroupby("col1")["col2"].describe() groups the data by the values found in the first column and then it focuses on the data found in column 2. The last part .describe() prints out the count, mean, min, max, std, Q1, Q2 and Q3 of the data based on the groups formed from the values in column 1. In the code the data was grouped by gender as I used gender for col1 and the data printed is based on the groupings relations to the values found under the column sibsp which was used for col2.,
8.2 when I ommitted the code to delete the column with missing values the results for dfgroupby("col1")["col2"].describe() remained the same.

8.3a When I showed the error message to ChatGPT it gave me the solution and told me and importing libraries which was much easier to comprehend than Google's response. When I searched "NameError: name 'pd' is not defined" into Google I had to search through StackOverflow to find an answer.

8.3b When I changed the ending titanic.csv to titanics.csv ChatGPT pointed it out and told me I had a typo in the url while Google was not as helpful. When I changed the entire url to titanics.csv the Chatbot told me that I needed to include the rest of the url and gave examples which was much more helpful than Google as I had to sift through articles to find similar cases.

8.3c Instead of using titanic for the df i used titanic and asked the Chatbot about the error message. It pointed out that I had not defined the dataframe as titanics and that I should be consistent with the variable names chosen. When I pasted the error message into google I got various answers which made it harder to use compared to the chatbot.

8.3d When leaving off a parathesis in line 4 I asked the chatbot about the error message which was "SyntaxError: '(' was never closed" and it gave examples of what may be causing it and how to fix it. Google actually gave me more complex answers making it harder to track the solutions.

8.3e To mistype one of the methods I typed describe() as describle() and sent the following error message to chatgpt "AttributeError: 'SeriesGroupBy' object has no attribute 'describle'". ChatGPT assumed that I was meaning to use the describe() method and showed me that I had to implement that instead. Google on the other hand was not very specific and gave me examples of other problems much more complex than my situation.

8.3f When capitalizing the column name when using groupby I received this message "KeyError: 'Column not found: Sibsp'". With this chatgpt told me that python is case sensitive and that I should double check the name of the column. Google was a bit more helpful and I found the same information but it was still easier to use the chatbot.

8.3g When I removed the parenthesis from the line "titanic.groupby("sex")["sibsp"].describe()" I was given this as the error "NameError: name 'sibsp' is not defined". When I asked the ChatBot about this it told me that I may not have been formatting correctly which was very useful. When I asked Google it did not give a clear solution.

Overall it was much easier to use the Chatbot as it answers the queries in the context of your code while on Google you have to work based off other's mistakes.



Chatbot link for PostLecture HW: https://chatgpt.com/share/66e378d7-65c0-8008-aaa9-27d427e97e7f

9. Yes I have been using chatGPT, the textbook and the discussion board.